File: FlashCard/Models/FlashCard.swift
import Foundation
import CloudKit

struct FlashCard: Identifiable, Codable, Hashable {
    var id = UUID()
    var word: String
    var definition: String // Translation
    var example: String
    var deckIds: Set<UUID>
    var successCount: Int = 0
    var dateCreated: Date = Date()
    var lastModified: Date = Date()
    
    // CloudKit tracking (stored as string to avoid serialization issues)
    var cloudKitRecordName: String?
    
    // Learning statistics
    var timesShown: Int = 0
    var timesCorrect: Int = 0
    
    // Spaced Repetition System (SRS) fields
    var srsLevel: Int = 0 // 0 = new, 1-10 = learning levels
    var nextReviewDate: Date? // When this card should be reviewed next
    var consecutiveCorrect: Int = 0 // Streak of correct answers
    var consecutiveIncorrect: Int = 0 // Streak of incorrect answers
    var easeFactor: Double = 2.5 // Multiplier for interval (starts at 2.5)
    var lastReviewDate: Date? // When this card was last reviewed
    var totalReviews: Int = 0 // Total number of reviews (not just shown)
    
    // Additional grammatical fields
    var article: String = "" // "het" or "de" for nouns
    var plural: String = "" // Plural form for nouns
    var pastTense: String = "" // Past tense form
    var futureTense: String = "" // Future tense form
    var pastParticiple: String = "" // Past participle form
    
    // Computed property for learning percentage
    var learningPercentage: Int? {
        guard timesShown > 0 else { return nil }
        return Int((Double(timesCorrect) / Double(timesShown)) * 100)
    }
    
    // Check if card is fully learned (5+ correct answers)
    var isFullyLearned: Bool {
        return timesCorrect >= 5
    }
    
    // MARK: - SRS Computed Properties
    
    // Check if card is due for review
    var isDueForReview: Bool {
        guard let nextReview = nextReviewDate else { return true }
        return Date() >= nextReview
    }
    
    // Check if card is new (never reviewed)
    var isNew: Bool {
        return srsLevel == 0 && totalReviews == 0
    }
    
    // Check if card is in learning phase (levels 1-3)
    var isLearning: Bool {
        return srsLevel >= 1 && srsLevel <= 3
    }
    
    // Check if card is in review phase (levels 4+)
    var isReviewing: Bool {
        return srsLevel >= 4
    }
    
    // Get the current interval in days
    var currentInterval: Int {
        switch srsLevel {
        case 0: return 0 // New card
        case 1: return 1 // 1 day
        case 2: return 6 // 6 days
        case 3: return 15 // 15 days
        default: return Int(pow(easeFactor, Double(srsLevel - 3)).rounded())
        }
    }
    
    // Get days until next review
    var daysUntilReview: Int? {
        guard let nextReview = nextReviewDate else { return nil }
        let calendar = Calendar.current
        let now = Date()
        return calendar.dateComponents([.day], from: now, to: nextReview).day
    }
    
    init(word: String = "", definition: String = "", example: String = "", deckIds: Set<UUID> = [], article: String = "", plural: String = "", pastTense: String = "", futureTense: String = "", pastParticiple: String = "", cardId: UUID? = nil, dateCreated: Date? = nil) {
        if let cardId = cardId {
            self.id = cardId
        }
        self.word = word
        self.definition = definition
        self.example = example
        self.deckIds = deckIds
        self.article = article
        self.plural = plural
        self.pastTense = pastTense
        self.futureTense = futureTense
        self.pastParticiple = pastParticiple
        self.successCount = 0
        self.timesShown = 0
        self.timesCorrect = 0
        self.dateCreated = dateCreated ?? Date()
        self.lastModified = Date()
    }
    
    // Migration initializer to handle old cards
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        id = try container.decode(UUID.self, forKey: .id)
        word = try container.decode(String.self, forKey: .word)
        definition = try container.decode(String.self, forKey: .definition)
        example = try container.decode(String.self, forKey: .example)
        deckIds = try container.decode(Set<UUID>.self, forKey: .deckIds)
        successCount = try container.decodeIfPresent(Int.self, forKey: .successCount) ?? 0
        dateCreated = try container.decodeIfPresent(Date.self, forKey: .dateCreated) ?? Date()
        lastModified = try container.decodeIfPresent(Date.self, forKey: .lastModified) ?? Date()
        timesShown = try container.decodeIfPresent(Int.self, forKey: .timesShown) ?? 0
        timesCorrect = try container.decodeIfPresent(Int.self, forKey: .timesCorrect) ?? 0
        
        // CloudKit properties
        cloudKitRecordName = try container.decodeIfPresent(String.self, forKey: .cloudKitRecordName)
        
        // Handle new fields with defaults for backward compatibility
        article = try container.decodeIfPresent(String.self, forKey: .article) ?? ""
        plural = try container.decodeIfPresent(String.self, forKey: .plural) ?? ""
        pastTense = try container.decodeIfPresent(String.self, forKey: .pastTense) ?? ""
        futureTense = try container.decodeIfPresent(String.self, forKey: .futureTense) ?? ""
        pastParticiple = try container.decodeIfPresent(String.self, forKey: .pastParticiple) ?? ""
        
        // SRS fields with defaults for backward compatibility
        srsLevel = try container.decodeIfPresent(Int.self, forKey: .srsLevel) ?? 0
        nextReviewDate = try container.decodeIfPresent(Date.self, forKey: .nextReviewDate)
        consecutiveCorrect = try container.decodeIfPresent(Int.self, forKey: .consecutiveCorrect) ?? 0
        consecutiveIncorrect = try container.decodeIfPresent(Int.self, forKey: .consecutiveIncorrect) ?? 0
        easeFactor = try container.decodeIfPresent(Double.self, forKey: .easeFactor) ?? 2.5
        lastReviewDate = try container.decodeIfPresent(Date.self, forKey: .lastReviewDate)
        totalReviews = try container.decodeIfPresent(Int.self, forKey: .totalReviews) ?? 0
    }
    
    private enum CodingKeys: String, CodingKey {
        case id, word, definition, example, deckIds, successCount, dateCreated, lastModified
        case timesShown, timesCorrect, article, plural, pastTense, futureTense, pastParticiple
        case cloudKitRecordName, srsLevel, nextReviewDate, consecutiveCorrect, consecutiveIncorrect
        case easeFactor, lastReviewDate, totalReviews
    }
    
    // Implement Hashable
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
    
    static func == (lhs: FlashCard, rhs: FlashCard) -> Bool {
        lhs.id == rhs.id
    }
    
    // MARK: - CloudKit Support
    
    static let recordType = "FlashCard"
    
    // Convert to CloudKit record
    func toCKRecord() -> CKRecord {
        let record: CKRecord
        
        if let recordName = cloudKitRecordName {
            let recordID = CKRecord.ID(recordName: recordName)
            record = CKRecord(recordType: Self.recordType, recordID: recordID)
        } else {
            record = CKRecord(recordType: Self.recordType)
        }
        
        record["id"] = id.uuidString
        record["word"] = word
        record["definition"] = definition
        record["example"] = example
        
        // Only set deckIds if it's not empty to avoid CloudKit empty array error
        let deckIdStrings = deckIds.map { $0.uuidString }
        if !deckIdStrings.isEmpty {
            record["deckIds"] = deckIdStrings
        }
        
        record["successCount"] = successCount
        record["dateCreated"] = dateCreated
        record["lastModified"] = lastModified
        record["timesShown"] = timesShown
        record["timesCorrect"] = timesCorrect
        record["article"] = article
        record["plural"] = plural
        record["pastTense"] = pastTense
        record["futureTense"] = futureTense
        record["pastParticiple"] = pastParticiple
        
        // SRS fields
        record["srsLevel"] = srsLevel
        if let nextReviewDate = nextReviewDate {
            record["nextReviewDate"] = nextReviewDate
        }
        record["consecutiveCorrect"] = consecutiveCorrect
        record["consecutiveIncorrect"] = consecutiveIncorrect
        record["easeFactor"] = easeFactor
        if let lastReviewDate = lastReviewDate {
            record["lastReviewDate"] = lastReviewDate
        }
        record["totalReviews"] = totalReviews
        
        return record
    }
    
    // Create from CloudKit record
    static func fromCKRecord(_ record: CKRecord) -> FlashCard? {
        guard let idString = record["id"] as? String,
              let id = UUID(uuidString: idString),
              let word = record["word"] as? String,
              let definition = record["definition"] as? String,
              let example = record["example"] as? String else {
            return nil
        }
        
        var card = FlashCard(
            word: word,
            definition: definition,
            example: example,
            cardId: id
        )
        
        // Restore deck IDs (field might not exist if it was empty)
        if let deckIdStrings = record["deckIds"] as? [String] {
            card.deckIds = Set(deckIdStrings.compactMap { UUID(uuidString: $0) })
        } else {
            card.deckIds = [] // Default to empty set if field doesn't exist
        }
        
        // Restore other properties
        card.successCount = record["successCount"] as? Int ?? 0
        card.dateCreated = record["dateCreated"] as? Date ?? Date()
        card.lastModified = record["lastModified"] as? Date ?? Date()
        card.timesShown = record["timesShown"] as? Int ?? 0
        card.timesCorrect = record["timesCorrect"] as? Int ?? 0
        card.article = record["article"] as? String ?? ""
        card.plural = record["plural"] as? String ?? ""
        card.pastTense = record["pastTense"] as? String ?? ""
        card.futureTense = record["futureTense"] as? String ?? ""
        card.pastParticiple = record["pastParticiple"] as? String ?? ""
        
        // SRS fields
        card.srsLevel = record["srsLevel"] as? Int ?? 0
        card.nextReviewDate = record["nextReviewDate"] as? Date
        card.consecutiveCorrect = record["consecutiveCorrect"] as? Int ?? 0
        card.consecutiveIncorrect = record["consecutiveIncorrect"] as? Int ?? 0
        card.easeFactor = record["easeFactor"] as? Double ?? 2.5
        card.lastReviewDate = record["lastReviewDate"] as? Date
        card.totalReviews = record["totalReviews"] as? Int ?? 0
        
        // Store CloudKit metadata
        card.cloudKitRecordName = record.recordID.recordName
        
        return card
    }
    
    // Update lastModified when card changes
    mutating func markAsModified() {
        lastModified = Date()
    }
} 

File: flashcard_app/lib/models/flash_card.dart
import 'dart:math';
import 'package:uuid/uuid.dart';

class FlashCard {
  final String id;
  String word;
  String definition; // Translation
  String example;
  Set<String> deckIds;
  int successCount;
  DateTime dateCreated;
  DateTime lastModified;
  
  // CloudKit tracking
  String? cloudKitRecordName;
  
  // Learning statistics
  int timesShown;
  int timesCorrect;
  
  // Spaced Repetition System (SRS) fields
  int srsLevel; // 0 = new, 1-10 = learning levels
  DateTime? nextReviewDate; // When this card should be reviewed next
  int consecutiveCorrect; // Streak of correct answers
  int consecutiveIncorrect; // Streak of incorrect answers
  double easeFactor; // Multiplier for interval (starts at 2.5)
  DateTime? lastReviewDate; // When this card was last reviewed
  int totalReviews; // Total number of reviews (not just shown)
  
  // Additional grammatical fields
  String article; // "het" or "de" for nouns
  String plural; // Plural form for nouns
  String pastTense; // Past tense form
  String futureTense; // Future tense form
  String pastParticiple; // Past participle form
  
  FlashCard({
    String? id,
    required this.word,
    String? definition,
    String? example,
    Set<String>? deckIds,
    this.successCount = 0,
    DateTime? dateCreated,
    DateTime? lastModified,
    this.cloudKitRecordName,
    this.timesShown = 0,
    this.timesCorrect = 0,
    this.srsLevel = 0,
    this.nextReviewDate,
    this.consecutiveCorrect = 0,
    this.consecutiveIncorrect = 0,
    this.easeFactor = 2.5,
    this.lastReviewDate,
    this.totalReviews = 0,
    this.article = '',
    this.plural = '',
    this.pastTense = '',
    this.futureTense = '',
    this.pastParticiple = '',
  }) : 
    id = id ?? const Uuid().v4(),
    definition = definition ?? '',
    example = example ?? '',
    deckIds = deckIds ?? {},
    dateCreated = dateCreated ?? DateTime.now(),
    lastModified = lastModified ?? DateTime.now();
  
  // Computed property for learning percentage
  int get learningPercentage {
    // Start with accuracy-based percentage
    if (timesShown == 0) return 0; // New cards start at 0%
    
    double basePercentage = (timesCorrect / timesShown) * 100;
    
    // Apply time decay if not recently reviewed
    double decayFactor = _calculateDecayFactor();
    
    return (basePercentage * decayFactor).clamp(0, 100).round();
  }
  
  // Calculate decay factor based on time since last review
  double _calculateDecayFactor() {
    if (lastReviewDate == null) return 1.0;
    
    final daysSinceReview = DateTime.now().difference(lastReviewDate!).inDays;
    
    // No decay for first 7 days
    if (daysSinceReview <= 7) return 1.0;
    
    // Gradual decay after 7 days: 5% per week
    final weeksSinceReview = (daysSinceReview - 7) / 7.0;
    final decayFactor = 1.0 - (weeksSinceReview * 0.05);
    
    return decayFactor.clamp(0.3, 1.0); // Minimum 30% retention
  }
  
  // Check if card is fully learned (5+ correct answers)
  bool get isFullyLearned {
    return timesCorrect >= 5;
  }
  
  // MARK: - SRS Computed Properties
  
  // Check if card is due for review
  bool get isDueForReview {
    if (nextReviewDate == null) return true;
    return DateTime.now().isAfter(nextReviewDate!);
  }
  
  // Check if card is new (never reviewed)
  bool get isNew {
    return srsLevel == 0 && totalReviews == 0;
  }
  
  // Check if card is in learning phase (levels 1-3)
  bool get isLearning {
    return srsLevel >= 1 && srsLevel <= 3;
  }
  
  // Check if card is in review phase (levels 4+)
  bool get isReviewing {
    return srsLevel >= 4;
  }
  
  // Get the current interval in days
  int get currentInterval {
    switch (srsLevel) {
      case 0: return 0; // New card
      case 1: return 1; // 1 day
      case 2: return 6; // 6 days
      case 3: return 15; // 15 days
      default: return (pow(easeFactor, srsLevel - 3)).round();
    }
  }
  
  // Get days until next review
  int? get daysUntilReview {
    if (nextReviewDate == null) return null;
    final now = DateTime.now();
    return nextReviewDate!.difference(now).inDays;
  }
  
  // MARK: - SRS Methods
  
  void markCorrect() {
    timesShown++;
    timesCorrect++;
    consecutiveCorrect++;
    consecutiveIncorrect = 0;
    totalReviews++;
    lastReviewDate = DateTime.now();
    lastModified = DateTime.now();
    
    // Update SRS level
    if (srsLevel < 10) {
      srsLevel++;
    }
    
    // Calculate next review date
    final interval = currentInterval;
    nextReviewDate = DateTime.now().add(Duration(days: interval));
    
    // Update ease factor (simplified algorithm)
    if (consecutiveCorrect >= 3) {
      easeFactor = (easeFactor + 0.1).clamp(1.3, 2.5);
    }
  }
  
  void markIncorrect() {
    timesShown++;
    consecutiveIncorrect++;
    consecutiveCorrect = 0;
    totalReviews++;
    lastReviewDate = DateTime.now();
    lastModified = DateTime.now();
    
    // Reset SRS level to 1 if it was higher
    if (srsLevel > 1) {
      srsLevel = 1;
    }
    
    // Calculate next review date (1 day for incorrect)
    nextReviewDate = DateTime.now().add(const Duration(days: 1));
    
    // Decrease ease factor
    easeFactor = (easeFactor - 0.2).clamp(1.3, 2.5);
  }
  
  // MARK: - JSON Serialization
  
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'word': word,
      'definition': definition,
      'example': example,
      'deckIds': deckIds.toList(),
      'successCount': successCount,
      'dateCreated': dateCreated.toIso8601String(),
      'lastModified': lastModified.toIso8601String(),
      'cloudKitRecordName': cloudKitRecordName,
      'timesShown': timesShown,
      'timesCorrect': timesCorrect,
      'srsLevel': srsLevel,
      'nextReviewDate': nextReviewDate?.toIso8601String(),
      'consecutiveCorrect': consecutiveCorrect,
      'consecutiveIncorrect': consecutiveIncorrect,
      'easeFactor': easeFactor,
      'lastReviewDate': lastReviewDate?.toIso8601String(),
      'totalReviews': totalReviews,
      'article': article,
      'plural': plural,
      'pastTense': pastTense,
      'futureTense': futureTense,
      'pastParticiple': pastParticiple,
    };
  }
  
  factory FlashCard.fromJson(Map<String, dynamic> json) {
    return FlashCard(
      id: json['id'],
      word: json['word'] ?? '',
      definition: json['definition'] ?? '',
      example: json['example'] ?? '',
      deckIds: Set<String>.from(json['deckIds'] ?? []),
      successCount: json['successCount'] ?? 0,
      dateCreated: json['dateCreated'] != null 
          ? DateTime.parse(json['dateCreated']) 
          : DateTime.now(),
      lastModified: json['lastModified'] != null 
          ? DateTime.parse(json['lastModified']) 
          : DateTime.now(),
      cloudKitRecordName: json['cloudKitRecordName'],
      timesShown: json['timesShown'] ?? 0,
      timesCorrect: json['timesCorrect'] ?? 0,
      srsLevel: json['srsLevel'] ?? 0,
      nextReviewDate: json['nextReviewDate'] != null 
          ? DateTime.parse(json['nextReviewDate']) 
          : null,
      consecutiveCorrect: json['consecutiveCorrect'] ?? 0,
      consecutiveIncorrect: json['consecutiveIncorrect'] ?? 0,
      easeFactor: (json['easeFactor'] ?? 2.5).toDouble(),
      lastReviewDate: json['lastReviewDate'] != null 
          ? DateTime.parse(json['lastReviewDate']) 
          : null,
      totalReviews: json['totalReviews'] ?? 0,
      article: json['article'] ?? '',
      plural: json['plural'] ?? '',
      pastTense: json['pastTense'] ?? '',
      futureTense: json['futureTense'] ?? '',
      pastParticiple: json['pastParticiple'] ?? '',
    );
  }
  
  // MARK: - Copy Methods
  
  FlashCard copyWith({
    String? word,
    String? definition,
    String? example,
    Set<String>? deckIds,
    int? successCount,
    DateTime? dateCreated,
    DateTime? lastModified,
    String? cloudKitRecordName,
    int? timesShown,
    int? timesCorrect,
    int? srsLevel,
    DateTime? nextReviewDate,
    int? consecutiveCorrect,
    int? consecutiveIncorrect,
    double? easeFactor,
    DateTime? lastReviewDate,
    int? totalReviews,
    String? article,
    String? plural,
    String? pastTense,
    String? futureTense,
    String? pastParticiple,
  }) {
    return FlashCard(
      id: id,
      word: word ?? this.word,
      definition: definition ?? this.definition,
      example: example ?? this.example,
      deckIds: deckIds ?? this.deckIds,
      successCount: successCount ?? this.successCount,
      dateCreated: dateCreated ?? this.dateCreated,
      lastModified: lastModified ?? this.lastModified,
      cloudKitRecordName: cloudKitRecordName ?? this.cloudKitRecordName,
      timesShown: timesShown ?? this.timesShown,
      timesCorrect: timesCorrect ?? this.timesCorrect,
      srsLevel: srsLevel ?? this.srsLevel,
      nextReviewDate: nextReviewDate ?? this.nextReviewDate,
      consecutiveCorrect: consecutiveCorrect ?? this.consecutiveCorrect,
      consecutiveIncorrect: consecutiveIncorrect ?? this.consecutiveIncorrect,
      easeFactor: easeFactor ?? this.easeFactor,
      lastReviewDate: lastReviewDate ?? this.lastReviewDate,
      totalReviews: totalReviews ?? this.totalReviews,
      article: article ?? this.article,
      plural: plural ?? this.plural,
      pastTense: pastTense ?? this.pastTense,
      futureTense: futureTense ?? this.futureTense,
      pastParticiple: pastParticiple ?? this.pastParticiple,
    );
  }
  
  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is FlashCard && other.id == id;
  }
  
  @override
  int get hashCode => id.hashCode;
  
  @override
  String toString() {
    return 'FlashCard(id: $id, word: $word, definition: $definition)';
  }
} 